### PR TITLE
Cast numeric to decimal256

### LINKED ISSUE #2922: Cast numeric to decimal256
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Increasing cast kernel coverage for Decimal256 by casting numeric to decimal256.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2922.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
@tustvold would it be possible to review this PR (so we can get it into the release candidate I plan to make tomorrow)?

**Review by tustvold — COMMENTED** [review]
I have a couple of reservations about merging this as is:

* The decimal256 conversion will silently wrap at the boundary of an i128
* The conversion in general is unchecked, which is inconsistent with the other numeric conversions

The former I think should be fixed before merge, the latter was already inconsistent for decimal128 so can probably slide

File: arrow/src/compute/kernels/cast.rs [hunk 1]
```diff
@@ -336,10 +356,25 @@ where
...
+
+fn cast_integer_to_decimal256<T: ArrowNumericType>(
+    array: &PrimitiveArray<T>,
+    precision: u8,
+    scale: u8,
+) -> Result<ArrayRef>
+where
+    <T as ArrowPrimitiveType>::Native: AsPrimitive<i256>,
+{
+    let mul: i256 = i256::from_i128(10_i128.pow(scale as u32));
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:370]: Can `10_i128.pow(DECIMAL256_MAX_SCALE as _)` overflow? `DECIMAL256_MAX_SCALE` is only 76.

[the remark this answers is not in the document]
Comment (tustvold) [reply arrow/src/compute/kernels/cast.rs:370]: Yes, https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2075e33b76d4894200dfe15d2a2cc37b

I suspect the max is DECIMAL128_MAX_SCALE which is 38

[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:370]: Ah, I missed this.

[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:370]: Add pow to i256: #2955

File: arrow/src/compute/kernels/cast.rs [hunk 2]
```diff
@@ -336,10 +356,25 @@ where
...
+
+    // with_precision_and_scale validates the
+    // value is within range for the output precision
+    cast_primitive_to_decimal256(array, |v| v.as_().wrapping_mul(mul), precision, scale)
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/compute/kernels/cast.rs:374]: We can fix decimal128 together later.

**Review by tustvold — APPROVED** [review]

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**